Assorted CLI tweaks

fredrikekelund opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Fixes #

## Proposed Changes

- Type tweak in `src/stores/snapshot-slice.ts`
- Added a standalone `getCliPath` function
- Removed refetch call for `useGetSnapshotUsage` query in `UserSettings` (see https://github.com/Automattic/studio/pull/1244#discussion_r2057697355)
- Added `--avoid-telemetry` option to CLI for when CLI is initiated internally from Studio

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Code review should suffice

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

fredrikekelund committed:
Type tweak

fredrikekelund committed:
Standalone getCliPath function

fredrikekelund committed:
Don't refetch useGetSnapshotUsage from component

fredrikekelund committed:
Added --avoid-telemetry option to CLI

fredrikekelund committed:
Fix tests

bcotrim commented on cli/index.ts:
```diff
@@ -31,6 +28,19 @@ async function main() {
 				return value;
 			},
 		} )
+		.option( 'avoid-telemetry', {
```
Nice one!

bcotrim reviewed (changes requested):
The number of active sites now doesn't get updated when we use the CLI until we refresh the app.


https://github.com/user-attachments/assets/21af8d9a-f16e-462f-930b-063058a79309

fredrikekelund commented:
Ah, I see what you were getting at, @bcotrim… Give me a minute to see if I can address that

fredrikekelund commented:
And just to clarify: the previous solution also has a bug in that deleting all sites from the settings modal refetches `useGetSnapshotUsage` too soon (i.e. before the changes have taken effect on the server). The result is that the preview site counter isn't decremented after the sites have been deleted.

fredrikekelund committed:
Invalidate query in `user-data-updated` handler

fredrikekelund committed:
Optimistically update query

fredrikekelund commented:
This should be good for another look now, @bcotrim!

bcotrim reviewed (commented):
My user has 2 preview sites outside of Studio.
So if I create a preview site with CLI, while having the user settings open, the number of sites will be wrong.
Opening and closing the user settings fixes this issue.

I believe it is an edge case and something we can live with for now, as I don't see a perfect solution. What do you think?

fredrikekelund committed:
Optimistically update query with diff count

The back-end site count can differ from the local site count. We respect this fact by using an increment operation to optimistically update the query instead of the absolute length of the snapshots array.

fredrikekelund commented:
Fair point, @bcotrim. I've updated the optimistic update to use an increment operation based on the difference between the old snapshot count and the new one. This accounts for scenarios like yours.

nightnei commented on cli/index.ts:
```diff
@@ -31,6 +28,19 @@ async function main() {
+			hidden: true,
+		} )
+		.middleware( async ( argv ) => {
+			if ( ! argv.avoidTelemetry ) {
```
Do we need this condition? AFAIU `avoidTelemetry` is always `true`? Or I misunderstand something?

fredrikekelund replied:
It's not true when the CLI is initiated directly from the terminal

nightnei reviewed (approved):
LGTM 👍
Left onr NIT comment
